File: openclaw/github_auth.js
const { chromium } = require('playwright');
const path = require('path');

(async () => {
  const userDataDir = path.join(process.env.HOME, '.openclaw', 'browser-profiles', 'openclaw');
  console.log('--- GITHUB DEVICE AUTHORIZATION ---');
  
  const browser = await chromium.launchPersistentContext(userDataDir, { headless: true });
  const page = await browser.newPage();
  
  try {
    console.log('Navigating to GitHub device login...');
    await page.goto('https://github.com/login/device', { waitUntil: 'load', timeout: 60000 });

    console.log('Detecting page state... current URL:', page.url());
    await page.screenshot({ path: 'github-debug-1.png' });

    if (page.url().includes('/login') && !page.url().includes('/device')) {
      console.log('ERROR: Redirected to login. Not authenticated.');
      process.exit(1);
    }

    console.log('Looking for user-code input...');
    const input = page.locator('#user-code');
    await input.waitFor({ state: 'visible', timeout: 5000 });
    
    await input.fill('837A-E26D');
    await page.keyboard.press('Enter');

    await page.waitForTimeout(3000);
    await page.screenshot({ path: 'github-debug-2.png' });

    const authorizeBtn = page.locator('button:has-text("Authorize"), button:has-text("Continue")');
    if (await authorizeBtn.isVisible()) {
      console.log('Clicking Authorize button...');
      await authorizeBtn.click();
      await page.waitForTimeout(3000);
      await page.screenshot({ path: 'github-debug-3.png' });
    }

    console.log('Final URL:', page.url());
    if (page.url().includes('success') || await page.isVisible('text=Congratulations')) {
      console.log('SUCCESS: Device authorized!');
      process.exit(0);
    } else {
      console.log('Verification failed.');
      process.exit(1);
    }

  } catch (err) {
    console.error('CRITICAL ERROR:', err.message);
    await page.screenshot({ path: 'github-error.png' });
    process.exit(1);
  } finally {
    await browser.close();
  }
})();

File: openclaw/github_auth.py
import asyncio
from playwright.async_api import async_playwright
import os

async def run_auth():
    print("--- GITHUB BROWSER AUTOMATION (PYTHON) ---")
    user_data_dir = os.path.expanduser("~/.openclaw/browser-profiles/openclaw")
    
    async with async_playwright() as p:
        # Launching with user data to hope for existing session
        context = await p.chromium.launch_persistent_context(
            user_data_dir,
            headless=True
        )
        page = await context.new_page()
        
        try:
            print("Navigating to https://github.com/login/device...")
            await page.goto("https://github.com/login/device", wait_until="networkidle", timeout=60000)
            
            # Check if login is needed
            if "login" in page.url and "device" not in page.url:
                print("!! REDIRECTED TO LOGIN. Looking for session cookies...")
                await page.screenshot(path="github_login_required.png")
                # Even if redirected, we'll try to find the code input just in case
            
            print(f"Current Title: {await page.title()}")
            print(f"Current URL: {page.url}")
            
            # Check for login form items
            if await page.locator('input[name="login"]').is_visible():
                print("!! CONFIRMED: ON LOGIN PAGE. Automation cannot proceed without manual login.")
                return 

            print("Looking for user-code input...")
            # Github uses a single input or multiple. Usually #user-code
            await page.wait_for_selector("#user-code", state="visible", timeout=10000)
            await page.fill("#user-code", "837A-E26D")
            print("Code entered. Submitting...")
            await page.keyboard.press("Enter")
            
            await asyncio.sleep(3)
            await page.screenshot(path="github_after_code.png")
            
            # Look for Authorize button
            auth_btn = page.locator('button:has-text("Authorize"), button:has-text("Continue")')
            if await auth_btn.is_visible():
                print("Clicking Authorize button...")
                await auth_btn.click()
                await asyncio.sleep(5)
                await page.screenshot(path="github_final.png")
            
            print(f"Final URL: {page.url}")
            if "success" in page.url.lower() or "congratulations" in await page.content():
                print("SUCCESS: Device Authorized!")
            else:
                print("Verification may have failed or requires manual login.")
                
        except Exception as e:
            print(f"ERROR: {str(e)}")
            await page.screenshot(path="github_error.png")
        finally:
            await context.close()

if __name__ == "__main__":
    asyncio.run(run_auth())
